src/main/java/dev/cdevents/models/TaskRunStartedSubject.java

```diff
@@ -120,6 +126,72 @@ public URI getSource() {
             public void setSource(URI source) {
                 this.source = source;
             }
+
+            /**
+             * @param o
+             * @return true or false
+             */
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) {
+                    return true;
+                }
+                if (!(o instanceof PipelineRun)) {
```

Comment:
Shouldn't this be `TaskRun` instead of `PipelineRun`? 
Since this is not caught by tests today, it looks like all cases where we use `equals` exit after the first check, or maybe they return `false` for the wrong reason.

It looks like we should have unit tests for this `equals` function specifically.
I'm a bit surprised that we need this though - this means that every optional field will require an `equals` function with dedicated logic and unit tests, and this adds a lot of complexity too when we will switch to generating the code.

If I read this logic correctly, this is what is implemented (pseudo-code):
```
aIdNoSource = {"id": "a" }
aIdNullSource =  {"id": "a", "source": null}
aIdSource = {"id": "a", "source": ""}
bIdNoSource = {"id": "a" }
bIdNullSource =  {"id": "a", "source": null}
bIdSource = {"id": "a", "source": ""}

aIdNoSource.equals(bIdNoSource)
aIdNullSource.equals(bIdNullSource)
aIdSource.equals(bIdSource)

all other combinations of "equals" are false.
```

I wonder which of these cases is not covered by the language default `equals` implementation?

Reply:
No, it is the Object of PipelineRun only we are using as per the spec,  https://github.com/cdevents/spec/blob/v0.3.0/core.md#taskrun-started

Actually the first check will return true only If we are passing the same object reference to the `equals` method, but from the test we are passing two different objects one is `expectedCDEvent` object from resource test data and the other created in the test with new object `cdEvent`.

`assertTrue(cdEvent.getSubject().getContent().equals(expectedCDEvent.getSubject().getContent()));
`

And this `equals` method will be covered fully I can debug when running this test itself and returns true as in this test all optional fields are null and all mandatory fields are set to valid and equal when comparing two objects.

Reply:
ok, thanks for the clarification.
If I understood it right, Java's default implementation of `equals` will only compare object references, so we need an implementation of `equals` that checks all properties one by one. Hopefully there is a way to generate a standard `equals` implementation in future.